apps/organizations/models.py

```diff
@@ -335,6 +337,87 @@ def build_notification_message__declined(self):
         return msg
 
 
+class BaseAbstractOrganizationJoinRequest(HtkBaseModel):
+    organization = fk_organization(related_name='join_requests', required=True)
+    user = fk_user(
+        related_name='organization_join_requests',
+    )
+    accepted = models.BooleanField(
+        default=None, null=True
+    )  # True: accepted, False: declined, None: not responded yet
+    timestamp = models.DateTimeField(auto_now_add=True)
+    responded_at = models.DateTimeField(blank=True, null=True, default=None)
+
+    class Meta:
+        abstract = True
+        verbose_name = 'Organization Join Request'
+
+    def __str__(self):
+        value = '{organization_name} - {user} - {status}'.format(
+            organization_name=self.organization.name,
+            user=self.user,
+            status=self.status,
+        )
+        return value
+
+    def json_encode(self) -> Dict[str, Any]:
+        """Returns a dictionary that can be `json.dumps()`-ed as a JSON representation of this object"""
+        value = {
+            'id': self.id,
+            'organization': self.organization.name,
+            'user': self.user.profile.get_full_name() if self.user else None,
+            'email': self.email,
+            'accepted': self.accepted,
+            'requested_at': self.timestamp,
+            'responded_at': self.responded_at,
+            'request_message': self.request_message,
+        }
+        return value
+
+    ##
+    # properties
+
+    @property
+    def status(self) -> str:
+        status = (
+            'Requested'
+            if self.accepted is None
+            else 'Accepted' if self.accepted else 'Declined'
+        )
+
+        return status
+
+    ##
+    # Notifications
+
+    def _build_notification_message(self, subject, action):
+        msg = '{subject_name} ({subject_username}<{email}>) request to join Organization <{organization_name}> has been {action}. Request Message - {message}'.format(  # noqa: E501
+            action=action,
+            subject_name=subject.profile.get_full_name(),
+            subject_username=subject.username,
+            email=subject.email,
+            organization_name=self.organization.name,
+            message=self.requested_message,
+        )
+        return msg
+
+    def build_notification_message__created(self):
+        msg = self._build_notification_message(self.user, 'sent')
+        return msg
+
+    def build_notification_message__resent(self):
+        msg = self._build_notification_message(self.user, 're-sent')
+        return msg
```

Comment:
Not sure if the re-sent state makes sense.

It makes sense for invitations, that you can re-send, because the insider wants the outsider to join, and the invitation may have gone missing.

For join requests, where the outsider is seeking to gain entrance to become an insider, does it make sense to allow re-request? Probably not? If I applied 1M times to Stanford University, this would seem excessive and harassing their admissions office.

Reply:
Right, that makes sense. What if there is an expiration for a request to join and then after that expiration time has ended, the user can make the request to join again. In this case, we would just use the normal request notification. I will remove this.